src/Functions/generateRandomStructure.cpp

```diff
@@ -0,0 +1,379 @@
+#include <Columns/ColumnString.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeFixedString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Interpreters/Context.h>
+#include <Common/randomSeed.h>
+
+#include <pcg_random.hpp>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+}
+
+class FunctionGenerateRandomStructure : public IFunction
+{
+private:
+    static constexpr std::array<TypeIndex, 29> simple_types
+    {
+        TypeIndex::Int8,
+        TypeIndex::UInt8,
+        TypeIndex::Int16,
+        TypeIndex::UInt16,
+        TypeIndex::Int32,
+        TypeIndex::UInt32,
+        TypeIndex::Int64,
+        TypeIndex::UInt64,
+        TypeIndex::Int128,
+        TypeIndex::UInt128,
+        TypeIndex::Int256,
+        TypeIndex::UInt256,
+        TypeIndex::Float32,
+        TypeIndex::Float64,
+        TypeIndex::Decimal32,
+        TypeIndex::Decimal64,
+        TypeIndex::Decimal128,
+        TypeIndex::Decimal256,
+        TypeIndex::Date,
+        TypeIndex::Date32,
+        TypeIndex::DateTime,
+        TypeIndex::DateTime64,
+        TypeIndex::String,
+        TypeIndex::FixedString,
+        TypeIndex::Enum8,
+        TypeIndex::Enum16,
+        TypeIndex::IPv4,
+        TypeIndex::IPv6,
+        TypeIndex::UUID,
+    };
+
+    static constexpr std::array<TypeIndex, 5> complex_types
+    {
+        TypeIndex::Nullable,
+        TypeIndex::LowCardinality,
+        TypeIndex::Array,
+        TypeIndex::Tuple,
+        TypeIndex::Map,
+    };
+
+    static constexpr std::array<TypeIndex, 19> map_key_types
+    {
+        TypeIndex::Int8,
+        TypeIndex::UInt8,
+        TypeIndex::Int16,
+        TypeIndex::UInt16,
+        TypeIndex::Int32,
+        TypeIndex::UInt32,
+        TypeIndex::Int64,
+        TypeIndex::UInt64,
+        TypeIndex::Int128,
+        TypeIndex::UInt128,
+        TypeIndex::Int256,
+        TypeIndex::UInt256,
+        TypeIndex::Date,
+        TypeIndex::Date32,
+        TypeIndex::DateTime,
+        TypeIndex::String,
+        TypeIndex::FixedString,
+        TypeIndex::IPv4,
+        TypeIndex::UUID,
+    };
+
+    static constexpr size_t MAX_NUMBER_OF_COLUMNS = 128;
+    static constexpr size_t MAX_TUPLE_ELEMENTS = 16;
+    static constexpr size_t MAX_DATETIME64_PRECISION = 9;
+    static constexpr size_t MAX_DECIMAL32_PRECISION = 9;
+    static constexpr size_t MAX_DECIMAL64_PRECISION = 18;
+    static constexpr size_t MAX_DECIMAL128_PRECISION = 38;
+    static constexpr size_t MAX_DECIMAL256_PRECISION = 76;
+    static constexpr size_t MAX_DEPTH = 32;
```

Comment:
Unused. Did you mean to set `allow_complex_types` to false if depth == MAX_DEPTH? (Making it a runtime argument instead of template, I guess.)

Reply:
Yeah, lost it while code rewriting. Thanks, returned it back.